File: src/main/java/at/yawk/fiction/impl/fanfiction/SubCategoryPageParser.java
package at.yawk.fiction.impl.fanfiction;

import at.yawk.fiction.impl.PageParser;
import java.util.ArrayList;
import java.util.List;
import org.jsoup.nodes.Element;

/**
 * @author yawkat
 */
class SubCategoryPageParser extends PageParser<List<FfnSubCategory>> {
    @Override
    protected List<FfnSubCategory> create() {
        return new ArrayList<>();
    }

    @Override
    protected void parse(Element root, List<FfnSubCategory> target) throws Exception {
        for (Element entry : root.select("#list_output tr td > div")) {
            FfnSubCategory subCategory = new FfnSubCategory();
            subCategory.setName(entry.select("a").text());
            String countText = extractGroup(entry.select("span").text(), "\\((\\d+(\\.\\d+)?[KM]?)\\)");
            subCategory.setEstimatedStoryCount(parseUnitedDouble(countText));
            target.add(subCategory);
        }
    }

    private static int parseUnitedDouble(String text) {
        char unit = text.charAt(text.length() - 1);
        if (Character.isDigit(unit)) {
            return Integer.parseInt(text);
        } else {
            int mul;
            switch (unit) {
            case 'K':
                mul = 1_000;
                break;
            case 'M':
                mul = 1_000_000;
                break;
            default:
                throw new UnsupportedOperationException("Unsupported unit '" + unit + "'");
            }
            return Math.round(Float.parseFloat(text.substring(0, text.length() - 1)) * mul);
        }
    }
}

File: src/main/kotlin/at/yawk/fiction/impl/fanfiction/SubCategoryPageParser.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package at.yawk.fiction.impl.fanfiction

import at.yawk.fiction.impl.PageParser
import org.jsoup.nodes.Element

/**
 * @author yawkat
 */
internal class SubCategoryPageParser(val category: FfnCategory) : PageParser<List<FfnSubCategory>>() {
    override fun parse(root: Element): List<FfnSubCategory> {
        return root.select("#list_output tr td > div").map {
            FfnSubCategory(
                    category = category,
                    name = it.select("a").text(),
                    estimatedStoryCount = parseUnitedDouble(extractGroup(
                            it.select("span").text(), "\\((\\d+(\\.\\d+)?[KM]?)\\)") ?: "")
            )
        }.toList()
    }

    private fun parseUnitedDouble(text: String): Int {
        val unit = text[text.length - 1]
        if (Character.isDigit(unit)) {
            return Integer.parseInt(text)
        } else {
            val mul: Int
            when (unit) {
                'K' -> mul = 1000
                'M' -> mul = 1000000
                else -> throw UnsupportedOperationException("Unsupported unit '$unit'")
            }
            return Math.round(java.lang.Float.parseFloat(text.substring(0, text.length - 1)) * mul)
        }
    }
}
